### PR TITLE
Make integration test able to generate a jUnit report 

### PR DESCRIPTION
**Pull request by jhadvig**
By adding starting the integration tests with `JUNIT_REPORT` variable, all tests will be run in and produce `intergration-test.log`, which in the end will be used by `junitreport` tool to generate a jUnit report - `report.xml`  

@stevekuznetsov PTAL

### CODE REVIEW

**Comment by jhadvig** [comment]
Also I've figured out that the coverage is not necessary for producing the jUnit report, so I we dont need to recompile the `integration.test` binary

File: hack/test-integration.sh [hunk 1]
```diff
@@ -96,16 +101,37 @@ loop="${TIMES:-1}"
...
+test_result="ok"
 pushd "${OS_ROOT}/${package}" &>/dev/null
+TESTSTARTTIME=$(date +%s%3N)
 for test in "${tests[@]}"; do
-	for((i=0;i<${loop};i+=1)); do
-		if ! (exectest "${test}" ${@:2}); then
-			ret=1
-		fi
-	done
+        for((i=0;i<${loop};i+=1)); do
```
Comment (stevekuznetsov) [inline hack/test-integration.sh]: Was this just indentation? Please put in another commit for easier review

File: hack/test-integration.sh [hunk 2]
```diff
@@ -96,16 +101,37 @@ loop="${TIMES:-1}"
...
+                        test_result="FAIL"
+                fi
+        done
 done
+TESTENDTIME=$(date +%s%3N)
+TESTDURATION=$(($TESTENDTIME - $TESTSTARTTIME))
+test_result_status="${test_result}        github.com/openshift/origin/test/integration    $(($TESTDURATION / 1000)).$(($TESTDURATION % 1000))s"
+if [[ -n "${test_output_file}" ]]; then
+	echo "${test_result_status}" >> "${test_output_file}"
+	cat ${test_output_file} | junitreport --type gotest --suites nested --stream --roots "github.com/openshift/origin/test/integration" --output "${junit_report_file}"
```
Comment (stevekuznetsov) [inline hack/test-integration.sh]: I would refactor `os::test::junit::generate_oscmd_report` to call something like `os::test::junit::generate_report` and write a new function `os::test::junit::generate_gotest_report` that you would trap just like we trap the `oscmd` stuff. Would be useful to share that with `hack/test-go.sh` as well.

**Comment by jhadvig** [comment]
No we dont need to add the `JUNIT_REPORT` env since we already provide it when running the tests in https://github.com/openshift/aos-cd-jobs/blob/master/sjb/config/test_cases/test_branch_origin_integration.yml#L10

File: hack/test-integration.sh (hunk 1 above)
Comment (jhadvig) [inline hack/test-integration.sh]: fixed... my IDE screwed me over :-/

**Comment by jhadvig** [comment]
@stevekuznetsov I've updated the PR.
Created the `os::test::junit::generate_report` function which takes the `--type` as an argument and is called by `os::test::junit::generate_oscmd_report`. 
Also refactored the `test-integration.sh` and `test-go.sh` to use the `os::test::junit::generate_report` function. Didnt wanted to create `os::test::junit::generate_gotest_report` function, since we are not using other functions like `os::test::junit::reconcile_output`, `os::test::junit::check_test_counters`, etc. and both  `test-integration.sh` and `test-go.sh` were just calling the junitreport tool

File: hack/test-integration.sh [hunk 3]
```diff
@@ -96,16 +101,39 @@ loop="${TIMES:-1}"
 # hack/test-integration.sh Template*
 # hack/test-integration.sh "(WatchBuilds|Template)"
 tests=( $(go run "${OS_ROOT}/hack/listtests.go" -prefix="${OS_GO_PACKAGE}/${package}.Test" "${testexec}" | grep -E "${1-Test}") )
+
+if [[ -n "${junit_report}" ]]; then
```
Comment (stevekuznetsov) [inline hack/test-integration.sh]: The convention is to use `$JUNIT_REPORT` to set these like: https://github.com/openshift/origin/blob/master/hack/test-cmd.sh#L79-L82

File: hack/test-integration.sh [hunk 4]
```diff
@@ -96,16 +101,39 @@ loop="${TIMES:-1}"
...
+	tmp_dir="/tmp/openshift/test-integration"
+	mkdir -p ${tmp_dir}
```
Comment (stevekuznetsov) [inline hack/test-integration.sh]: Just use pre-existing temp directories:

```sh
export JUNIT_REPORT_OUTPUT="${LOG_DIR}/raw_test_output.log"
```

File: hack/test-integration.sh [hunk 5]
```diff
@@ -96,16 +101,39 @@ loop="${TIMES:-1}"
...
+	test_output_file="${tmp_dir}/test-go.log"
+	test_error_file="${tmp_dir}/test-go-err.log"
```
Comment (stevekuznetsov) [inline hack/test-integration.sh]: We're not using this for anything in here -- if you are trying to replicate the checks we have in `hack/test-go.sh` for panics, etc -- factor those, too, into common functions.

File: hack/test-integration.sh [hunk 6]
```diff
@@ -96,16 +101,39 @@ loop="${TIMES:-1}"
...
+	junit_report_file="${tmp_dir}/report.xml"
+fi
 # run each test as its own process
 ret=0
+test_result="ok"
 pushd "${OS_ROOT}/${package}" &>/dev/null
+os::log::info "Running \`go test\`..."
```
Comment (stevekuznetsov) [inline hack/test-integration.sh]: Not sure this a super useful info

File: hack/test-integration.sh [hunk 7]
```diff
@@ -96,16 +101,39 @@ loop="${TIMES:-1}"
...
 			ret=1
+			test_result="FAIL"
 		fi
 	done
 done
+TESTENDTIME=$(date +%s%3N)
+TESTDURATION=$(($TESTENDTIME - $TESTSTARTTIME))
+test_result_status="${test_result}        github.com/openshift/origin/test/integration    $(($TESTDURATION / 1000)).$(($TESTDURATION % 1000))s"
+if [[ -n "${test_output_file}" ]]; then
+	echo "${test_result_status}" | tee "${test_output_file}"
```
Comment (stevekuznetsov) [inline hack/test-integration.sh]: Pipe to `tee` is weird -- use `>>`

File: hack/test-integration.sh [hunk 8]
```diff
@@ -96,16 +101,39 @@ loop="${TIMES:-1}"
...
 for test in "${tests[@]}"; do
 	for((i=0;i<${loop};i+=1)); do
 		if ! (exectest "${test}" ${@:2}); then
 			ret=1
+			test_result="FAIL"
 		fi
 	done
 done
+TESTENDTIME=$(date +%s%3N)
+TESTDURATION=$(($TESTENDTIME - $TESTSTARTTIME))
```
Comment (stevekuznetsov) [inline hack/test-integration.sh]: Convention is use to use `$lower_case_variables` for those you're not exporting

File: hack/test-integration.sh [hunk 9]
```diff
@@ -96,16 +101,39 @@ loop="${TIMES:-1}"
...
+test_result_status="${test_result}        github.com/openshift/origin/test/integration    $(($TESTDURATION / 1000)).$(($TESTDURATION % 1000))s"
+if [[ -n "${test_output_file}" ]]; then
+	echo "${test_result_status}" | tee "${test_output_file}"
+	JUNIT_REPORT_OUTPUT="${test_output_file}" ARTIFACT_DIR="${tmp_dir}" os::test::junit::generate_report "gotest"
+else
+	echo "${test_result_status}"
```
Comment (stevekuznetsov) [inline hack/test-integration.sh]: Not sure this is a super useful message either -- if you set `$JUNIT_REPORT_OUTPUT` in the `if [[ -n "${JUNIT_REPORT:-}" ]]` stanza like we do for other tests, and use `$ARTIFACT_DIR` from the setup scripts, this could look like:

```sh
echo "${test_result}        github.com/openshift/origin/test/integration    $((test_duration / 1000)).$((test_duration % 1000))s" >> "${JUNIT_REPORT_OUTPUT:-/dev/null}"
os::test::junit::generate_report "gotest"
```

File: hack/test-integration.sh (hunk 5 above)
Comment (stevekuznetsov) [inline hack/test-integration.sh]: These would be helpful inside of a `os::test::junit::generate_gotest_report`

**Comment by jhadvig** [comment]
@stevekuznetsov comments addressed. Thanks !

File: hack/test-integration.sh [hunk 10]
```diff
@@ -96,16 +101,33 @@ loop="${TIMES:-1}"
 # hack/test-integration.sh Template*
 # hack/test-integration.sh "(WatchBuilds|Template)"
 tests=( $(go run "${OS_ROOT}/hack/listtests.go" -prefix="${OS_GO_PACKAGE}/${package}.Test" "${testexec}" | grep -E "${1-Test}") )
+
+if [[ -n "${JUNIT_REPORT:-}" ]]; then
```
Comment (stevekuznetsov) [inline hack/test-integration.sh]: Move this up to line 32 for visual clarity -- should execute fine but is weird that it's after the `exectest()` declaration.

File: hack/test-integration.sh [hunk 11]
```diff
@@ -96,16 +101,33 @@ loop="${TIMES:-1}"
...
 for test in "${tests[@]}"; do
 	for((i=0;i<${loop};i+=1)); do
 		if ! (exectest "${test}" ${@:2}); then
 			ret=1
+			test_result="FAIL"
 		fi
 	done
 done
+test_end_time=$(date +%s%3N)
+test_duration=$(($test_end_time - $test_start_time))
```
Comment (stevekuznetsov) [inline hack/test-integration.sh]: Don't need and should not use `$` for var expansion inside of `$(( ))`

**Comment by stevekuznetsov** [comment]
[test]

File: hack/lib/test/junit.sh [hunk 12]
```diff
@@ -172,10 +172,22 @@ function os::test::junit::generate_oscmd_report() {
     # use the junitreport tool to generate us a report
     os::util::ensure::built_binary_exists 'junitreport'
```
Comment (stevekuznetsov) [inline hack/lib/test/junit.sh]: @jhadvig this should be in `os::test::junit::generate_report()`

Comment (stevekuznetsov) [inline hack/lib/test/junit.sh]: Same with the `if` statement:

```sh
    if [[ -z "${JUNIT_REPORT_OUTPUT:-}" ||
          -n "${JUNIT_REPORT_OUTPUT:-}" && ! -s "${JUNIT_REPORT_OUTPUT:-}" ]]; then
          # we can't generate a report
          return
    fi
```

**Comment by jhadvig** [comment]
@stevekuznetsov thanks for comments and nits ! Should be fixed now

**Comment by stevekuznetsov** [comment]
Integation jUnit looks good:

https://ci.openshift.redhat.com/jenkins/job/test_pull_request_origin_integration/1773/testReport/github/com_openshift_origin_test_integration /

[merge]
